eyelid-match-ops/src/primitives/yashe.rs

```diff
@@ -121,17 +226,53 @@ where
             // This is ok because the C::Coeff modulus is smaller than MIN/MAX.
             res[i] = C::Coeff::from(v as i64);
         }
+
+        // Raw coefficient access must be followed by a truncation check.
         res.truncate_to_canonical_form();
+
         res
     }
 
     /// Sample a polynomial with unlimited size random coefficients using a uniform distribution.
-    pub fn sample_uniform(&self, mut rng: &mut ThreadRng) -> Poly<C> {
+    pub fn sample_uniform_coeff(&self, mut rng: &mut ThreadRng) -> Poly<C> {
         let mut res = Poly::non_canonical_zeroes(C::MAX_POLY_DEGREE);
         for i in 0..C::MAX_POLY_DEGREE {
             let coeff_rand = C::Coeff::rand(&mut rng);
             res[i] = coeff_rand;
         }
+
+        // Raw coefficient access must be followed by a truncation check.
+        res.truncate_to_canonical_form();
+        res
+    }
+
+    /// Sample from message space
+    pub fn sample_message(&self, rng: &mut ThreadRng) -> Message<C> {
+        let m = self.sample_uniform_range(0..C::T, rng);
+        Message { m }
+    }
+
+    /// Sample from binary message space
+    pub fn sample_binary_message(&self, rng: &mut ThreadRng) -> Message<C> {
+        // TODO: this might be implemented more efficiently using `Rng::gen_bool()`
+        let m = self.sample_uniform_range(0..=1_u64, rng);
+        Message { m }
+    }
+
+    /// Sample a polynomial with random coefficients in `range` using a uniform distribution.
+    fn sample_uniform_range<T, R>(&self, range: R, rng: &mut ThreadRng) -> Poly<C>
+    where
+        T: SampleUniform,
+        R: SampleRange<T> + Clone,
+        C::Coeff: From<T>,
+    {
+        let mut res = Poly::non_canonical_zeroes(C::MAX_POLY_DEGREE);
+        for i in 0..C::MAX_POLY_DEGREE {
+            let coeff_rand = rng.gen_range(range.clone());
+            res[i] = coeff_rand.into();
+        }
```

Comment:
We were repeating the same code in two different functions, so it seemed helpful to have a generic sampler for any coefficient range.